ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java

```diff
@@ -0,0 +1,777 @@
+   */
+  @Override
+  void handleResponse(ResponseInfo responseInfo) {
+    GetChunk getChunk =
+        correlationIdToGetChunk.remove(((RequestOrResponse) responseInfo.getRequest()).getCorrelationId());
+    getChunk.handleResponse(responseInfo);
+    if (getChunk.isComplete()) {
+      onChunkOperationComplete(getChunk);
+    }
+  }
+
+  /**
+   * Attempt to write the data associated with the blob to the channel passed in by the caller (if the caller has
+   * done so).
+   */
+  private void maybeWriteToChannel() {
+    // if there are chunks available to be written out, do now.
+    if (firstChunk.isComplete() && readCalled) {
+      while (indexOfNextChunkToWriteOut < numChunksTotal && chunkIndexToBuffer
+          .containsKey(indexOfNextChunkToWriteOut)) {
+        ByteBuffer chunkBuf = chunkIndexToBuffer.remove(indexOfNextChunkToWriteOut);
+        asyncWritableChannel.write(chunkBuf, chunkAsyncWriteCallback);
+        indexOfNextChunkToWriteOut++;
+      }
+      if (operationException.get() != null || numChunksWrittenOut.get() == numChunksTotal) {
+        completeRead();
+        operationCompleted = true;
+      }
+    }
+  }
+
+  /**
+   * Complete the read from this {@link ReadableStreamChannel} by invoking the callback and marking the future.
+   */
+  void completeRead() {
+    if (readIntoCallbackCalled.compareAndSet(false, true)) {
+      if (operationException.get() != null) {
+        bytesWritten = null;
+      }
+      readIntoFuture.done(bytesWritten, operationException.get());
+      if (readIntoCallback != null) {
+        readIntoCallback.onCompletion(bytesWritten, operationException.get());
+      }
+    }
+  }
+
+  /**
+   * As part of the poll, GetBlobOperation fetches new requests to be issued and updates the state based on
+   * whether previously issued requests have timed out and so on. Additionally, any writes of previously retrieved
+   * chunk buffers into the {@link AsyncWritableChannel} passed in by the caller also happens within this method.
+   * @param requestRegistrationCallback the {@link RequestRegistrationCallback} to call for every request that gets
+   *                                    created as part of this poll operation.
+   */
+  @Override
+  void poll(RequestRegistrationCallback<GetOperation> requestRegistrationCallback) {
+    if (operationCompleted) {
+      return;
+    }
+    if (firstChunk.isReady() || firstChunk.isInProgress()) {
+      firstChunk.poll(requestRegistrationCallback);
+    }
+    if (firstChunk.isComplete()) {
+      // Although an attempt is made to write to the channel as soon as a chunk is successfully retrieved,
+      // the caller might not have called readInto() and passed in a channel at the time. So an attempt is always
+      // made from within poll.
+      maybeWriteToChannel();
+      // If this is a composite blob, poll for requests for subsequent chunks.
+      if (dataChunks != null) {
+        for (GetChunk dataChunk : dataChunks) {
```

Comment:
Also, can you help me understand what happens in this case or is my understanding right. Lets say we poll for the data chunks here and have collected all the required requests that needs to be sent out, but just before returning an exception was thrown via chunkAsyncWriteCallback and we complete the operation. But then, we have registered all the requests with the requestRegistrationCallback which means that the NBR will send out those requests which is of no use. Only during next poll(), we won't revisit this operation again, coz GetManager.poll() would have removed this GetOperation from its list since the operation is complete.